File: Part 6 - Reinforcement Learning/Section 32 - Upper Confidence Bound (UCB)/my_ucb.py
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import math

# Importing the testing dataset
dataset = pd.read_csv('Ads_CTR_Optimisation.csv')

# Initialize variables, test set has 10 ads
N = len(dataset)
d = 10
selections = [0] * d
rewards = [0] * d
ads_selected = []
# Carry out test for every observation in test set
for n in range(N):
    # Init variables for choice information
    max_ucb_add = 0
    max_ucb = 0
    # Calculate upper bound for every ad
    # First ten iterations just show subsequent ads
    for i in range(d):
        if selections[i] > 0:
            avg_reward = rewards[i] / selections[i]
            delta_i = math.sqrt(3/2 * (math.log(n + 1)/selections[i]))
            ucb = avg_reward + delta_i
        else:
            ucb = 1e400
        if ucb > max_ucb:
            max_ucb = ucb
            max_ucb_add = i
    # Check chosen ad, add reward and selection information
    ads_selected.append(max_ucb_add)
    selections[max_ucb_add] += 1
    reward = dataset.iloc[n,max_ucb_add]
    rewards[max_ucb_add] += reward
    

File: Part 6 - Reinforcement Learning/Section 32 - Upper Confidence Bound (UCB)/my_ucb.R
# Importing the testing dataset
dataset = read.csv('Ads_CTR_Optimisation.csv')

# Initialize variables, test set has 10 ads
N = 10000
d = 10
ads_selected = c()
selections_per_ad = integer(d)
rewards_per_ad = integer(d)
# Carry out test for every observation in test set
for(n in 1:N){
  # Init variables for choice information
  max_ucb_add = 0
  max_ucb = 0
  # Calculate upper bound for every ad
  # First ten iterations just show subsequent ads
  for(i in 1:d){
    if(selections_per_ad[i] > 0){
      avg_ad_reward = rewards_per_ad[i] / selections_per_ad[i]
      delta_i = sqrt(3/2 * (log(n)/selections_per_ad[i]))
      ucb = avg_ad_reward + delta_i
    }else{
      ucb = 1e400
    }
    if(ucb > max_ucb){
      max_ucb = ucb
      max_ucb_add = i
    }
  }
  # Check chosen ad, add reward and selection information
  reward = dataset[n,max_ucb_add]
  rewards_per_ad[max_ucb_add] = rewards_per_ad[max_ucb_add] + reward
  selections_per_ad[max_ucb_add] = selections_per_ad[max_ucb_add] + 1
  ads_selected = append(ads_selected, max_ucb_add)
}

# Check freqeuncy for each ad
hist(ads_selected,
     ylab = "Count",
     xlab = "Ads",
     main = "Selections per ad")
